# Read .mat data in github

link of the code and data: https://github.com/pbashivan/EEGLearn

Data for Visual WM experiment described in:
Bashivan, Pouya, et al. "Learning Representations from EEG with Deep Recurrent-Convolutional Neural Networks." International conference on learning representations (2016).

FeatureMat_timeWin:
FFT power values extracted for three frequency bands (theta, alpha, beta). Features are arranged in band and electrodes order (theta_1, theta_2..., theta_64, alpha_1, alpha_2, ..., beta_64). There are seven time windows, features for each time window are aggregated sequentially (i.e. 0:191 --> time window 1, 192:383 --> time windw 2 and so on. Last column contains the class labels (load levels).

Neuroscan_locs_orig:
3 dimensional coordinates for electrodes on Neuroscan quik-cap.

trials_subNums:
contains subject numbers associated with each trial (used for leave-subject-out cross validation).

In [19]:
import scipy.io as scio
import numpy as np
import pandas as pd 
 

data_dict = scio.loadmat('./data/FeatureMat_timeWin.mat')
coordinates_dict = scio.loadmat('./data/Neuroscan_locs_orig.mat')

In [20]:
data_dict.keys()
coordinates_dict.keys()

dict_keys(['__header__', '__version__', '__globals__', 'A'])

In [21]:
coordinates = coordinates_dict['A']
df_features = data_dict['features']

In [23]:
df_features.shape
# coordinates.shape

(2670, 1345)

In [26]:
df_features[1]

array([ 1.15085451,  1.02301335,  0.48282339, ...,  0.59169703,
       -0.01846558,  1.        ])

So far, we know that, their data has 64 signal channel, each electrode has 192 feature(alpha beta theta * 64), and each data has 7 time windows, so in total 1344 data + 1 label for one line. and this data has 2670 small set for train;

We can also find the head plot function in OpenBCI GUI, is kind of like this process.


![head_electrode_position](./data/img_notebook/Cyton_16Ch_Static_Headplot_Image.png)

The above paper is not the only way that we could test, and because of data size limitations, we may consider data augmentation methods!

# Connection test 

The connection with Hugo has been very successful, he did some test for the motion imagination, so for me, I will start from object imagination!

And for how to prepare the equipment for data collections and do the experiments:

1. Please read https://docs.openbci.com/AddOns/Headwear/GelfreeElectrodeCap/
2. For the document in 1, I also made a brief summary after reading it. You can also read my summary to get the environment configuration and other related information. https://xulintaofr.gitbook.io/eeg/

(This gitbook was originally designed to organize all the practice content during my internship, but due to time constraints, only the configuration of the device and the functional details of the equipment were summarized.)

## Experimental procedure

0. If the electrodes are not connected, please follow this picture and connect them.

![OpenBCI_channels](./data/img_notebook/channels.jpg)

1. Get the device and check everything is right;

2. Make the saline solution by dissolving 1 tea spoon of sodium chloride (~6 g) into about 200 mL tap water with the cup. Soak the Hydro-linkTM( the white things) sponges in the saline solution for 5-10 min;

3. Make sure you don't put the saline solution at another table, do not put them with our board in same place!!!!!!!!!!

4. Connect USB to the computer, switch ''OFF'' to ''PC'' in our board, open the OpenBCI GUI application;

5. In the application page, chick the " CYTON(live)", then click "Serial(from Dongle)", click "Manual" to choose our port, then choose "16 CHANNELS"; You can change the setting - SESSION DATA Name and Max File Duration,then click "Start session" and waiting your signals.

![OpenBCI_GUI](./data/img_notebook/GUI.jpg)

And for Europe we should change to Notch to 50Hz!

6. Insert the Hydro-linkTM into the electrode holder of the electrode lead wire.

7. Place the cap on the subject's head. Make sure the Cz is correctly located halfway on the midline of the head.

8. Do your own experiments, you have better not wear the cap more than 30 minutes/ 1 hour.

9. Turn off your GUI application and switch ''PC'' to ''OFF''

10. After the measurement is finished, disconnect the electrode connectors from the CytonDaisy board. Remove the electrode holders out of the bases on the cap. 1) Take the Hydro-linkTM sponges out of the holders and wash the Hydro-linkTM with tap water for at least three times, while squeezing out the water in the Hydro-linkTM during washing. Then let the Hydro-linkTM dry in air. 2) Rinse the electrode holders and the Gelfree BCI cap with tap water. 3) Ensure that everything is dry before storing in a dry place away from sunlight

11. Cleaning frequency-after each use.

12. If you won't use the cap in a long time, please disconnect the electrodes and clean the cap.



PS: you can find your recoding txt files in C:\Users\XXX(i.e.Lintao XU)\Documents\OpenBCI_GUI

# Experiment object

For my experiments, I prefer to do the experiments like the OpenBCI and create experiment' videos to help control the timing of the experiment and determine the signal interval.

(Files all in the folder my experiment!)

Objects I will use:

## ASL
* American Sign language Letters
1. Vision (observe the letters)
2. Imaginations (imagine the letters)

## AAT
* Alphabet Letters
1. Vision (observe the letters)
2. Imaginations (imagine the letters)

The big problem that we have is the dataset size! 
For DL we need data for model training!


May consider the public dataset!
But remember the initial idea for our project is, create the device for one person, not for all person.

(in fact no such data set is open!...)

# Experiment design

The codes for creating videos in the rep ''my_experiment'', are the official codes of OpenBCI(rep-OpenBCI_official_experiment), but I modify the codes for our needs.

### REP(folders) FILES

* alphabet_sounds: One person read all alphabets letters, and modified audio files.

* asl_piece: Some pieces videos for asl, make sure delete all files before you run code video_audio_asl.py.

* experiments_videos: All experiments' videos for asl and alphabets.

* Images: Original images and new named images, and welcome/video beginning images.

* labels_alphabets: Labels order for the videos generated by the file video_creator_alphabet.py, and these labels will be used by the file video_audio_asl.py.

### Code files

* video_images_prepare.py: Randomly select 5(defined) images of each type from the data of the two original image datasets, and copy and rename them to another folder(Letter_class and Sign_class).

<code>python video_images_prepare.py</code>

* audio_add.py: Code(ffmpeg) to extend audio, merge audio with video, concat videos!

* settings.json: Default setting, not modify them unless you have to.

* join.txt: File for ffmpeg to merge two videos, automatic modified by codes.

* video_creator_alphabet.py: Generate a fixed-length alphabets video, modify the video name and txt name in the main function.

<code>python video_creator_alphabet.py -1 settings.json</code>

* video_audio_asl.py: Generate a fixed-length asl video by the given alphabet_labels order with audio for each sign letter, modify the video name and given alphabet_labels txt name in the main function.

<code>python video_audio_asl.py -1 settings.json</code>

* video_creator_asl.py: Generate a fixed-length asl video, modify the video name and txt name in the main function. This file no longer use!!!

<code>python video_creator_asl.py -1 settings.json</code>

(If you are reading this notebook and have some questions after you have finished my codes reading, please contact me.)

And for the video asl, I took the advice from prof, I added the voice to read ''A, B, C...'', but I really think it's better to remember the ASL instead of add voice to the video, because I think it'll be harder when you want to think the ASL in the imagination part(you could easily directly think of alphabet letters rather than sign letters)

For the video design,

for now, in each videos, Images will show 3.5 seconds, and then you need to think/imagine the images 3 seconds.

The video is so long, so keep patients when do the experiments!

Notice:

For some unknown reasons, I missed letter "X" in alphabet videos, besides, because the video pieces are generates in 

2.5 sec black + 3 sec images + 0.5 sec black

so the last image of each video will lack time to imagine it.

So I do another video and add 30 times letter X and 6 last missed letters(4-Y, 2-G) and other letters A-Z twice. 

And the videos for asl are kept same with alphabets labels order and videos!

(PS: After I found I missed time of imagination for last letter, I fixed this problem in the codes, X videos, so if re run these codes, videos will have time to imagine the last letters)

So in total, I have signal data for ASL and Alphabet:

* 6 videos(each one have 125 letters and 124 have both imaginations and visions)
* X extensional video(30 X, 2*26 A-Z, 4 Y, 2 G, totally 88 data)

For the other future video design ways.

1. Make the each image time longer than 3 seconds etc

2. Just voice and no images, closed eyes and imagine our letters.

3. For the ASL part, need to re consider.